watchtower/blob/justice_kit_test.go

```diff
@@ -212,3 +216,195 @@ func testBlobJusticeKitEncryptDecrypt(t *testing.T, test descriptorTest) {
+	if err != blob.ErrNoCommitToRemoteOutput {
+		t.Fatalf("expected ErrNoCommitToRemoteOutput, got: %v", err)
+	}
+}
+
+// TestJusticeKitToLocalWitnessConstruction tests that a JusticeKit returns the
+// proper to-local witness script and to-local witness stack for spending the
+// revocation path.
+func TestJusticeKitToLocalWitnessConstruction(t *testing.T) {
+	csvDelay := uint32(144)
+
+	// Generate the revocation and delay private keys.
+	revPrivKey, err := btcec.NewPrivateKey(btcec.S256())
+	if err != nil {
+		t.Fatalf("unable to generate revocation priv key: %v", err)
+	}
+
+	delayPrivKey, err := btcec.NewPrivateKey(btcec.S256())
+	if err != nil {
+		t.Fatalf("unable to generate delay priv key: %v", err)
+	}
+
+	// Copy the revocation and delay pubkeys into the format expected by our
+	// justice kit.
+	var revPubKey blob.PubKey
+	copy(revPubKey[:], revPrivKey.PubKey().SerializeCompressed())
+
+	var delayPubKey blob.PubKey
+	copy(delayPubKey[:], delayPrivKey.PubKey().SerializeCompressed())
+
+	// Sign a message using the revocation private key. The exact message
+	// doesn't matter as we won't be validating the signature's validity.
+	digest := bytes.Repeat([]byte("a"), 32)
+	rawRevSig, err := revPrivKey.Sign(digest)
+	if err != nil {
+		t.Fatalf("unable to generate revocation signature: %v", err)
+	}
+
+	// Convert the DER-encoded signature into a fixed-size sig.
+	commitToLocalSig, err := lnwire.NewSigFromSignature(rawRevSig)
+	if err != nil {
+		t.Fatalf("unable to convert raw revocation signature to "+
+			"Sig: %v", err)
+	}
+
+	// Populate the justice kit with fields relevant to the to-local output.
+	justiceKit := &blob.JusticeKit{
+		CSVDelay:         csvDelay,
+		RevocationPubKey: revPubKey,
+		LocalDelayPubKey: delayPubKey,
+		CommitToLocalSig: commitToLocalSig,
+	}
+
+	// Compute the expected to-local script, which is a function of the CSV
+	// delay, revocation pubkey and delay pubkey.
+	expToLocalScript, err := lnwallet.CommitScriptToSelf(
+		csvDelay, delayPrivKey.PubKey(), revPrivKey.PubKey(),
+	)
+	if err != nil {
+		t.Fatalf("unable to generate expected to-local script: %v", err)
+	}
+
+	// Compute the to-local script that is returned by the justice kit.
+	toLocalScript, err := justiceKit.CommitToLocalWitnessScript()
+	if err != nil {
+		t.Fatalf("unable to compute to-local witness script: %v", err)
+	}
+
+	// Assert that the expected to-local script matches the actual script.
+	if !bytes.Equal(expToLocalScript, toLocalScript) {
+		t.Fatalf("mismatched to-local witness script, want: %v, got %v",
+			expToLocalScript, toLocalScript)
+	}
+
+	// Next, compute the to-local witness stack returned by the justice kit.
+	toLocalWitnessStack, err := justiceKit.CommitToLocalRevokeWitnessStack()
+	if err != nil {
+		t.Fatalf("unable to compute to-local witness stack: %v", err)
+	}
+
+	// A valid witness that spends the revocation path should have exactly
+	// two elements on the stack.
+	if len(toLocalWitnessStack) != 2 {
```

Comment:
The revoke witness stack should have 3 items: the sig, the 0x01, and the witness script.

Reply:
i've been using witness stack here to describe everything excluding the witness script, which the caller already has via `CommitToLocal/RemoteWitnessScript()`. they are later joined via: https://github.com/lightningnetwork/lnd/blob/master/watchtower/lookout/justice_descriptor.go#L293